src/ir/auth_logic/universe_relation_insertion.cc

```diff
@@ -0,0 +1,357 @@
+      : decl_env_(decl_env) {}
+
+  // The `AddUniverseConditions` methods, including this one,
+  // are not implemented as actual visits because these methods
+  // are implementing a transformation, and the returned values
+  // need to be placed in a part of the syntactic context that
+  // is specific to each syntactic element.
+  SaysAssertion AddUniverseConditions(const SaysAssertion& says_assertion) {
+    std::vector<Assertion> assertion_list;
+    for (const Assertion& assertion : says_assertion.assertions()) {
+      assertion_list.push_back(AddUniverseConditions(assertion));
+    }
+    return SaysAssertion(says_assertion.principal(), std::move(assertion_list));
+  }
+
+ private:
+  Assertion AddUniverseConditions(const Assertion& assertion) {
+    return std::visit(
+        [this](const auto& variant) { return AddUniverseConditions(variant); },
+        assertion.GetValue());
+  }
+
+  Assertion AddUniverseConditions(const Fact& fact) {
+    std::vector<BaseFact> conditions = GetUniverseConditions(fact.base_fact());
+    if (conditions.size() > 0) {
+      return Assertion(ConditionalAssertion(fact, {conditions}));
+    } else {
+      return Assertion(fact);
+    }
+  }
+
+  Assertion AddUniverseConditions(const ConditionalAssertion& cond_assertion) {
+    std::vector<BaseFact> conditions =
+        GetUniverseConditions(cond_assertion.lhs().base_fact());
+    absl::c_copy(cond_assertion.rhs(), std::back_inserter(conditions));
+    return Assertion(ConditionalAssertion(cond_assertion.lhs(), conditions));
+  }
+
+  // This method, given a BaseFact appearing in the LHS of an assertion,
+  // returns a vector of conditions that should be added to the RHS of that
+  // assertion. This method is where this class is actually used as a visitor.
+  // The visitor works only on subtrees of the AST that are
+  // these BaseFacts on the LHS and returns a set of BaseFacts with
+  // the universe conditions.
+  inline std::vector<BaseFact> GetUniverseConditions(
+      const BaseFact& base_fact) {
+    return base_fact.Accept(*this);
+  }
+
+  std::vector<BaseFact> GetDefaultValue() override { return {}; }
+  std::vector<BaseFact> CombineResult(std::vector<BaseFact> left,
+                                      std::vector<BaseFact> right) override {
+    left.insert(left.end(), right.begin(), right.end());
+    return left;
+  }
+
+  std::vector<BaseFact> PreVisit(const Principal& prin) override {
+    if (!IsNameConstant(prin.name())) {
+      return {MakeUniverseMembershipFact(
+          prin.name(), datalog::ArgumentType::MakePrincipalType())};
+    } else {
+      return {};
+    }
+  }
+
+  // Add universe conditions for the non-constant arguments
+  std::vector<BaseFact> Visit(const datalog::Predicate& pred) override {
+    datalog::RelationDeclaration decl =
+        decl_env_.GetDeclarationOrFatal(pred.name());
+    std::vector<BaseFact> conditions;
+    for (unsigned i = 0; i < pred.args().size(); i++) {
```

Comment:
Use `size_t` rather than `unsigned`, as on a 64 bit system, `unsigned` isn't wide enough to cover the range of potential indices in `pred.args().size()`.